# Task 4.8

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

## Importing df and checking data

In [2]:
path = r'D:\Downloads\Data Analytics\content\Task 4'

In [3]:
combined_df_47 = pd.read_pickle(os.path.join(path, 'Data', 'combined_df_47.pkl'))

In [4]:
combined_df_47.shape

(32404859, 18)

In [5]:
combined_df_47.head()

order_id  user_id  order_number  orders_day_of_week  time_of_order  \
0   2539329        1             1                   2              8   
1   2398795        1             2                   3              7   
2    473747        1             3                   3             12   
3   2254736        1             4                   4              7   
4    431534        1             5                   4             15   

   days_without_orders  product_id  add_to_cart_order  reordered product_name  \
0                  0.0         196                  1          0         Soda   
1                 15.0         196                  1          1         Soda   
2                 21.0         196                  1          1         Soda   
3                 29.0         196                  1          1         Soda   
4                 28.0         196                  1          1         Soda   

   aisle_id  department_id  prices _merge    price_range_loc     busiest_day  \
0        77              7     9.0   both  Mid-range product  Regularly busy   
1        77              7     9.0   both  Mid-range product  Regularly busy   
2        77              7     9.0   both  Mid-range product  Regularly busy   
3        77              7     9.0   both  Mid-range product      Least busy   
4        77              7     9.0   both  Mid-range product      Least busy   

     busiest_days busiest_period_of_day  
0  Regularly busy        Average orders  
1    Slowest days        Average orders  
2    Slowest days           Most orders  
3    Slowest days        Average orders  
4    Slowest days           Most orders

## 4.8.2 and 4.8.3

In [6]:
combined_df_47.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                15.457838   1  99
2                17.277920   1  99
3                17.170395   1  99
4                17.811403   1  99
5                15.215751   1  99
6                16.439806   1  99
7                17.225802   1  99
8                15.340650   1  99
9                15.895474   1  99
10               20.197148   1  99
11               16.170638   1  99
12               15.887671   1  99
13               16.583536   1  99
14               16.773669   1  99
15               16.165037   1  99
16               17.665606   1  99
17               15.694469   1  99
18               19.310397   1  99
19               17.177343   1  99
20               16.473447   1  99
21               22.902379   1  99

In this scenario we can see the importance of having a complete dataset instead of just a portion.
For example at **Department:17**, we can see an increase in the mean from **11.294069** to **15.694469**. We can also notice that some departments, like **21** have the highest mean, which was not available on the subset, meaning that the department with the highest mean changed from **Department:19 with a mean of 19.305237**, to **Department:21 with a mean of 22.902379**.

## 4.8.4

In [7]:
# We start by creating a new collumn that we will call 'max_order'
combined_df_47['max_order'] = combined_df_47.groupby(['user_id'])['order_number'].transform(np.max)

In [8]:
# Then we test it to confirm it.
combined_df_47.head(60)

order_id  user_id  order_number  orders_day_of_week  time_of_order  \
0    2539329        1             1                   2              8   
1    2398795        1             2                   3              7   
2     473747        1             3                   3             12   
3    2254736        1             4                   4              7   
4     431534        1             5                   4             15   
5    3367565        1             6                   2              7   
6     550135        1             7                   1              9   
7    3108588        1             8                   1             14   
8    2295261        1             9                   1             16   
9    2550362        1            10                   4              8   
10   2968173       15            15                   1              9   
11   1870022       15            17                   2             16   
12   1911383       15            18                   2             11   
13   2715276       15            21                   1              9   
14    487368       15            22                   1             10   
15   2293453       19             2                   5             14   
16   1973799       19             5                   6             12   
17    532817       19             7                   4             17   
18   1573906       21            10                   3             10   
19   1593000       31            10                   3              8   
20   2231262       31            17                   3             11   
21   2580647       43             6                   4             16   
22   2187180       43             9                   4             12   
23   2497897       52             1                   1              9   
24   1318871       52             2                   1             10   
25   1261384       52             3                   1             10   
26    580568       52             5                   2             10   
27   2428073       52             7                   2              7   
28    342306       52             9                   1             11   
29   1520353       52            10                   2             10   
30   2542086       52            11                   2              9   
31    944694       52            12                   1             16   
32   1268191       52            15                   2             11   
33   1122089       52            17                   2             10   
34   1498922       52            18                   2             10   
35    180919       52            19                   3              8   
36   2695875       52            21                   5             10   
37   3205960       67             1                   3             11   
38    970922       67             2                   1             15   
39    596585       67             3                   1             13   
40   2037087       67             4                   2              9   
41   2474111       67             5                   3             14   
42   1568481       67             6                   3             12   
43   2048527       67             8                   2             11   
44   2389031       67             9                   2             14   
45   2378562       67            10                   1             10   
46   1515698       67            11                   1             11   
47   1928050       67            12                   2             12   
48   1949274       67            13                   1             10   
49    551869       67            14                   1              9   
50    768572       67            16                   4             10   
51    150528       67            18                   5             16   
52     29758       67            19                   0             14   
53   155

In [9]:
# Finnaly we use loc() to create the loyalty flags
combined_df_47.loc[combined_df_47['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [10]:
combined_df_47.loc[(combined_df_47['max_order'] <= 40) & (combined_df_47['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [11]:
combined_df_47.loc[combined_df_47['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [12]:
combined_df_47['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

## 4.8.5

In [13]:
combined_df_47.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

As we can see, it does make sense that new customers tend to have a higher mean, as customers that buy frenquently mostly likely are re-purchasing products that they already use, while new customers are trying different products, which may result in more spending. 

## 4.8.6

In [14]:
# We start by creating a new column showing the mean of prices

In [15]:
combined_df_47['spend_mean'] = combined_df_47.groupby(['user_id'])['prices'].transform(np.mean)

In [16]:
combined_df_47

order_id  user_id  order_number  orders_day_of_week  time_of_order  \
0          2539329        1             1                   2              8   
1          2398795        1             2                   3              7   
2           473747        1             3                   3             12   
3          2254736        1             4                   4              7   
4           431534        1             5                   4             15   
...            ...      ...           ...                 ...            ...   
32404854   1320836   202557            17                   2             15   
32404855     31526   202557            18                   5             11   
32404856    758936   203436             1                   2              7   
32404857   2745165   203436             2                   3              5   
32404858   3093936   205420             1                   4             14   

          days_without_orders  product_id  add_to_cart_order  reordered  \
0                         0.0         196                  1          0   
1                        15.0         196                  1          1   
2                        21.0         196                  1          1   
3                        29.0         196                  1          1   
4                        28.0         196                  1          1   
...                       ...         ...                ...        ...   
32404854                  1.0       43553                  2          1   
32404855                  3.0       43553                  2          1   
32404856                  0.0       42338                  4          0   
32404857                 15.0       42338                 16          1   
32404858                  0.0       28818                  8          0   

                           product_name  ...  department_id  prices  _merge  \
0                                  Soda  ...              7     9.0    both   
1                                  Soda  ...              7     9.0    both   
2                                  Soda  ...              7     9.0    both   
3                                  Soda  ...              7     9.0    both   
4                                  Soda  ...              7     9.0    both   
...                                 ...  ...            ...     ...     ...   
32404854            Orange Energy Shots  ...              7     3.7    both   
32404855            Orange Energy Shots  ...              7     3.7    both   
32404856          Zucchini Chips, Pesto  ...             19     6.9    both   
32404857          Zucchini Chips, Pesto  ...             19     6.9    both   
32404858  Hot Oatmeal Multigrain Raisin  ...             14    10.3    both   

            price_range_loc     busiest_day    busiest_days  \
0         Mid-range product  Regularly busy  Regularly busy   
1         Mid-range product  Regularly busy    Slowest days   
2         Mid-range product  Regularly busy    Slowest days   
3         Mid-range product      Least busy    Slowest days   
4         Mid-range product      Least busy    Slowest days   
...                     ...             ...             ...   
32404854  Low-range product  Regularly busy  Regularly busy   
32404855  Low-range product  Regularly busy  Regularly busy   
32404856  Mid-range product  Regularly busy  Regularly busy   
32404857  Mid-range product  Regularly busy    Slowest days   
32404858  Mid-range product      Least busy    Slowest days   

         busiest_period_of_day max_order      loyalty_flag spend_mean  
0               Average orders        10      New customer   6.367797  
1               Average orders        10      New customer   6.367797  
2                  Most orders        10      New customer   6.367797  
3               Average orders        10      New customer   6.367797  
4                  Most orders        10      New customer   6.367797  
...       

In [17]:
# Then we create the conditions for the groups of High and low spender clients

In [18]:

combined_df_47.loc[combined_df_47['spend_mean'] > 10, 'spender_type'] = 'High spender'

In [19]:
combined_df_47.loc[combined_df_47['spend_mean'] <= 10, 'spender_type'] = 'Low spender'

In [20]:
# And finally we count them

In [21]:
combined_df_47['spender_type'].value_counts(dropna = False)

Low spender     31770995
High spender      633864
Name: spender_type, dtype: int64

## 4.8.7 

In [22]:
#Same as before, we start by creating a new column for the median of days_since_prior_order

In [23]:
combined_df_47['avg_days_last_order'] = combined_df_47.groupby(['user_id'])['days_without_orders'].transform(np.median)

In [24]:
# Then we create the groups

In [25]:
combined_df_47.loc[combined_df_47['avg_days_last_order'] > 20, 'customer_frequency'] = 'Non-frequent customer'

In [26]:
combined_df_47.loc[(combined_df_47['avg_days_last_order'] <= 20) & (combined_df_47['avg_days_last_order'] > 10), 'customer_frequency'] = 'Regular customer'

In [27]:
combined_df_47.loc[combined_df_47['avg_days_last_order'] <= 10, 'customer_frequency'] = 'Frequent customer'

In [28]:
# And finally count them again

In [29]:
combined_df_47['customer_frequency'].value_counts(dropna = False)

Frequent customer        22796659
Regular customer          6921472
Non-frequent customer     2686728
Name: customer_frequency, dtype: int64

## Exporting

In [30]:
combined_df_47.to_pickle(os.path.join(path, 'Data', 'combined_df_48.pkl'))